File: integrations/src/main/java/framework/core/service/ThirdPartyServiceConfigReadService.java
package framework.core.service;

import framework.config.ThirdPartyServiceList;
import framework.core.data.RunConfig;
import framework.core.exception.ServiceConfigurationException;
import framework.core.repository.ThirdPartyServiceConfigRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class ThirdPartyServiceConfigReadService {

    private final ThirdPartyServiceConfigRepository thirdPartyServiceConfigRepository;
    private final ThirdPartyProviderConfigReadService thirdPartyProviderConfigReadService;

    @Autowired
    public ThirdPartyServiceConfigReadService(
            ThirdPartyServiceConfigRepository thirdPartyServiceConfigRepository,
            ThirdPartyProviderConfigReadService thirdPartyProviderConfigReadService) {
        this.thirdPartyServiceConfigRepository = thirdPartyServiceConfigRepository;
        this.thirdPartyProviderConfigReadService = thirdPartyProviderConfigReadService;
    }

    public RunConfig findByService(ThirdPartyServiceList service) {
        var serviceEntity = thirdPartyServiceConfigRepository.findByServiceAndIsActiveTrue(service.getServiceName())
                .orElseThrow(() -> new ServiceConfigurationException("Error fetching " + service));

        var primaryConfig = thirdPartyProviderConfigReadService.getProviderConfigById(serviceEntity.getPrimaryConfigKey());
        var fallbackConfig = serviceEntity.getFallbackConfigKey() != null
                ? thirdPartyProviderConfigReadService.getProviderConfigById(serviceEntity.getFallbackConfigKey())
                : null;
        
        return new RunConfig(primaryConfig, fallbackConfig, serviceEntity.getRetryCount());
    }
}

File: integrations/src/main/kotlin/framework/core/service/ThirdPartyServiceConfigReadService.kt
package framework.core.service

import framework.config.ThirdPartyServiceList
import framework.core.data.RunConfig
import framework.core.exception.ServiceConfigurationException
import framework.core.repository.ThirdPartyServiceConfigRepository
import org.springframework.stereotype.Service

@Service
class ThirdPartyServiceConfigReadService(
    private val thirdPartyServiceConfigRepository: ThirdPartyServiceConfigRepository,
    private val thirdPartyProviderConfigReadService: ThirdPartyProviderConfigReadService
) {

    fun findByService(service: ThirdPartyServiceList): RunConfig {
        val serviceEntity = thirdPartyServiceConfigRepository.findByServiceAndIsActiveTrue(service.serviceName)
            ?: throw ServiceConfigurationException("Error fetching $service")

        val primaryConfig = thirdPartyProviderConfigReadService.getProviderConfigById(serviceEntity.primaryConfigKey)
        val fallbackConfig =
            serviceEntity.fallbackConfigKey?.let { thirdPartyProviderConfigReadService.getProviderConfigById(it) }
        return RunConfig(primaryConfig, fallbackConfig, serviceEntity.retryCount)
    }
}
